commit 24e932357ee3041763135b931206dfc0bbe0441e
Author: rnhmjoj <rnhmjoj@inventati.org>
Date:   Wed Jul 23 10:18:55 2025 +0200

    Fixes for running wpa_supplicant unprivileged
    
    1. Change the dbus service user to "wpa_supplicant"
    
    2. Ensure appropriate group ownership and permissions on the client sockets.
       Motivation: clients communicate with the daemon by creating "client"
       sockets; by default this is owned by the user running the client,
       so it may be inaccessible by the daemon.
    
    3. Move the "control" sockets under a subdirectory of /run/wpa_supplicant.
       Motivation: wpa_supplicant will try to adjust the ownership of the
       sockets directory, even if they are fine, and fail.
    
    4. Move the "client" under a subdirectory of /run/wpa_supplicant instead
       of tmp. Motivation: this allows to unshare /tmp

diff --git a/src/common/wpa_ctrl.c b/src/common/wpa_ctrl.c
index 7e197f0..6bfb091 100644
--- a/src/common/wpa_ctrl.c
+++ b/src/common/wpa_ctrl.c
@@ -15,6 +15,8 @@
 #include <fcntl.h>
 #include <sys/un.h>
 #include <unistd.h>
+#include <sys/types.h>
+#include <grp.h>
 #include <fcntl.h>
 #endif /* CONFIG_CTRL_IFACE_UNIX */
 #ifdef CONFIG_CTRL_IFACE_UDP_REMOTE
@@ -165,6 +167,14 @@ try_again:
 		return NULL;
 	}
 
+	/* Set the client socket owner group to "wpa_supplicant"
+	 * and ensure group and user permissions are the same */
+	struct group *grp = getgrnam("wpa_supplicant");
+	if (grp != NULL) {
+		lchown(ctrl->local.sun_path, -1, grp->gr_gid);
+		chmod(ctrl->local.sun_path, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
+	}
+
 #ifdef ANDROID
 	/* Set group even if we do not have privileges to change owner */
 	lchown(ctrl->local.sun_path, -1, AID_WIFI);
--- a/wpa_supplicant/dbus/dbus-wpa_supplicant.conf
+++ b/wpa_supplicant/dbus/dbus-wpa_supplicant.conf
@@ -2,9 +2,15 @@
  "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
  "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
 <busconfig>
-        <policy user="root">
+        <policy user="wpa_supplicant">
                 <allow own="fi.w1.wpa_supplicant1"/>
-
+        </policy>
+        <policy user="root">
+                <allow send_destination="fi.w1.wpa_supplicant1"/>
+                <allow send_interface="fi.w1.wpa_supplicant1"/>
+                <allow receive_sender="fi.w1.wpa_supplicant1" receive_type="signal"/>
+        </policy>
+        <policy group="wpa_supplicant">
                 <allow send_destination="fi.w1.wpa_supplicant1"/>
                 <allow send_interface="fi.w1.wpa_supplicant1"/>
                 <allow receive_sender="fi.w1.wpa_supplicant1" receive_type="signal"/>
diff --git a/wpa_supplicant/dbus/fi.w1.wpa_supplicant1.service.in b/wpa_supplicant/dbus/fi.w1.wpa_supplicant1.service.in
index d97ff39..367a7c6 100644
--- a/wpa_supplicant/dbus/fi.w1.wpa_supplicant1.service.in
+++ b/wpa_supplicant/dbus/fi.w1.wpa_supplicant1.service.in
@@ -1,5 +1,5 @@
 [D-BUS Service]
 Name=fi.w1.wpa_supplicant1
 Exec=@BINDIR@/wpa_supplicant -u
-User=root
+User=wpa_supplicant
 SystemdService=wpa_supplicant.service
diff --git a/wpa_supplicant/wpa_cli.c b/wpa_supplicant/wpa_cli.c
index af00e79..840b307 100644
--- a/wpa_supplicant/wpa_cli.c
+++ b/wpa_supplicant/wpa_cli.c
@@ -44,10 +44,10 @@ static int wpa_cli_attached = 0;
 static int wpa_cli_connected = -1;
 static int wpa_cli_last_id = 0;
 #ifndef CONFIG_CTRL_IFACE_DIR
-#define CONFIG_CTRL_IFACE_DIR "/var/run/wpa_supplicant"
+#define CONFIG_CTRL_IFACE_DIR "/run/wpa_supplicant/control"
 #endif /* CONFIG_CTRL_IFACE_DIR */
 static const char *ctrl_iface_dir = CONFIG_CTRL_IFACE_DIR;
-static const char *client_socket_dir = NULL;
+static const char *client_socket_dir = "/run/wpa_supplicant/client";
 static char *ctrl_ifname = NULL;
 static const char *global = NULL;
 static const char *pid_file = NULL;
